### DISCUSSION TITLE
[ARCHIVE] Oh god my map is conflicted: A guide on how to fix map merge conflicts.

### DISCUSSION THREAD

**Discussion by S34NW** [ParadiseSS13/Paradise#16402] [Guides]
THIS IS MASSIVELY OUTDATED, PLEASE HOLD WHILE A NEW GUIDE IS PUT TOGETHER

# Important note:
**This guide will NOT help you if your edits and the master map update the same tiles. In that case, there is no option but to manually compare both and keep all the appropriate changes from both maps.
Always make sure you have looked at the mapping PRs made since you started your project, they might touch on the same areas as you and you will have to apply their changes to your changes.**

**Failing to do this _will_ lead to you undoing changes and making more work for everyone down the line.**

## Preface
This guide works on the assumption of basic StrongDMM (SDMM) and Git knowledge, and the use of VSC with the GitHub extension.
I will attempt to avoid assuming you know all the terms and acronyms, and I will illustrate where possible with screenshots.

## Oh my god my map is conflicted send help!

We've all been there, and joining the mapping queue means you are expected to deconflict your own map when you are next in the mapping queue.
The first thing you will want to do is make a backup copy of your edited map. To do this: 

1. Swap to the branch with your edits within VSC.
2. Locate the map file, and _copy_ it to a folder outside of your Git workspace. (Perhaps in your user documents or, if you like a big mess, your desktop.)
3. Open GitHub and download a copy of the latest master version of the map you edit, save this in the same place as you backed up your edits.

Now that you have backed everything up, it's time to bring your branch up to date with master.
In VSC, do a merge master. This is done by swapping to _your_ branch and right-clicking master to get the correct option.
![2021-07-25 11_45_51-S34N - Visual Studio Code](https://user-images.githubusercontent.com/12197162/126896828-81c79618-b534-4cb7-87da-a012c5bee318.png)

In the top-left section of your window, you should now see something like this if you have done it correctly:
![2021-07-25 11_46_24-S34N - Visual Studio Code](https://user-images.githubusercontent.com/12197162/126896849-9d328723-f0c7-4967-a3a9-d88ead4a87aa.png)

Open the conflicted file by clicking it, give it a moment to figure out the changes in your file, then right-click on it to select this option: (It doesn't matter what option we select here, we will be overwriting the file regardless.)
![2021-07-25 11_49_00-delta dmm - S34N - Visual Studio Code](https://user-images.githubusercontent.com/12197162/126896892-087dd005-2e65-41bc-aca6-1065bc786e9b.png)

Once that has been completed, stage the map and commit your merge master. Then, go into your files and replace the map with the updated master map you downloaded earlier.

### This section applies to edits that fit in a rectangle. If your edits are a strange shape or are map-wide, there is an additional section below.

Once you have done this, open SDMM.
Load up the fresh, master map we just placed in the normal map folder.
Also, open your backed up map from wherever you stored it.
**OPEN BOTH MAPS IN THE SAME SDMM INSTANCE, YOU CAN THEN SWAP BETWEEN MAPS IN THE TOP RIGHT.**

Now, identify the areas that your edits fill. In my example edit, I have edited the bridge section of deltastation, including the surrounding hallways.
![2021-07-25 11_53_21-paradise  _delta dmm backup  - StrongDMM](https://user-images.githubusercontent.com/12197162/126897031-35c6be80-713c-471b-81f8-b94421c19567.png)

1. Select your changed area.
2. Copy it to clipboard.
3. Swap back to the "master" map and paste your edits. Your mouse should be in the bottom left of your changes.

That is it, you have applied your edits to the latest master map.
Swap back to VSC and you should see something similar to this:
![2021-07-25 11_54_50-S34N - Visual Studio Code](https://user-images.githubusercontent.com/12197162/126897128-abb4d44c-9e02-4a8c-acce-89c564d095b0.png)
Fill a commit name, and commit the changes.
And there you go, done!

### Edits to specific objects only.

Okay, let's say you have re-piped the entire map. (Or made changes to specific item types only.) How the hell do we copy changes over without re-writing everything else?

The answer is within SDMM's "Layers Filter" feature.
Open it from Options -> Layers Filter...
Lets say we want to change only atmospherics pipes and devices.

Untick everything barring atmospherics items.
![2021-07-25 11_57_24-paradise  _maps_map_files_Delta_delta dmm  - StrongDMM](https://user-images.githubusercontent.com/12197162/126897276-0a00e232-02fb-4ed8-9f75-ada83464e8ce.png)

Go to your master map, and with the filter applied select everything and delete it all.
Now go back to your backed up edits map. With the filter applied, copy the entire map, and paste it over the now-blank master.
This will insert only the item you have selected in the filters, in our case atmos pipes.

_Only_ things ticked in the "Layers Filter" menu will be copied. So you can safely apply just those without touching anything else.

Now you can save your edits and commit as above.

### Edits that are not a square

What if we want to make changes in an area that is not a square, without overwriting parts of the map we haven't touched?
We could just break it up into multiple copy/pastes, but this can be hard for irregular or awkward shapes. There is an easier way!

Let's say we have edited botany and the kitchen as in the below selection.
![2021-07-25 11_58_58-paradise  _maps_map_files_Delta_delta dmm  - StrongDMM](https://user-images.githubusercontent.com/12197162/126897379-f1434181-0712-40d8-ace3-aface5faacd2.png)

First, delete the area around that we have made no edits to, so it should look like this:
![2021-07-25 11_59_36-paradise  _maps_map_files_Delta_delta dmm  - StrongDMM](https://user-images.githubusercontent.com/12197162/126897400-90d9ac71-bbae-4b17-9dd2-53062043c3cc.png)

If we copy this as above, you notice it will place space when we try to paste it in. How do we avoid this?
Remove space from the layers filter:
![2021-07-25 11_59_47-paradise  _maps_map_files_Delta_delta dmm  - StrongDMM](https://user-images.githubusercontent.com/12197162/126897429-0711062b-1c23-40f8-a6b6-0c11327d4400.png)

By doing this, when we copy and paste these edits onto the master map, it will only place the turfs that we can see in the above screenshot.

Now you can save your edits and commit as above.

## Conclusion

After doing this, you will still have the merge conflict tag on GitHub. This is normal, the next time a maintainer merges a PR, it will get removed.

The above methods should give you enough options to easily deconflict any change. If you have made mass edits across the map to many different items, there is no easy solution and you just have to put in the effort to deconflict. You can always poke the #coding_chat channel on discord for more help if needed.

A final note, please test your changes after updating like this. There could be unintended interactions with other changes on the map and you should always re-test your edits.

**Comment by FreeStylaLT** [comment]
![img](https://cdn.discordapp.com/emojis/842257327566487572.png?v=1)
![image](https://user-images.githubusercontent.com/11361525/126897746-d4c4dd67-92f3-4924-84a3-75dcc0398803.png)
